### PR TITLE
[WOR-431] Refactor V2 billing project creation code

### PR DESCRIPTION
**Pull request by aherbst-broad**
## Context
We need to be able to create azure billing profiles in billing profile manager as part of azure billing project creation. This PR is the first step in the refactoring necessary to introduce that functionality. This step is focused solely on refactoring of existing functionality with an eye toward review-ability and more easily digestible PRs; it does **not** introduce any new azure/bpm related code.

## This PR
* Extracts the V2 billing project creation code path to a separate class for improved testability. The current coverage is achieved via heavyweight integration tests on the V2 billing endpoints. I've dubbed this class `BillingProjectOrchestrator`.
* Introduces a billing project repository, again for improved testability. The intention of this class is to isolate all slick DB calls for billing project operations and allow the orchestration layer to test it's business logic without the somewhat elaborate mocking needed for slick data sources or spinning up a real DB connection.

## Next Steps
* Introduce the Azure managed application billing profile creation code path. This will be implemented as a follow-on PR, currently [up in draft](https://github.com/broadinstitute/rawls/pull/1961)
------------
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by aherbst-broad** [comment]
jenkins retest

File: core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingRepository.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.broadinstitute.dsde.rawls.dataaccess.SlickDataSource
+import org.broadinstitute.dsde.rawls.model.{RawlsBillingProject, RawlsBillingProjectName}
+
+import scala.concurrent.Future
+
+/**
+ * Data access for rawls billing projects
+ */
+class BillingRepository(dataSource: SlickDataSource) {
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingRepository.scala:11]: I added this class to improve testability of the `BillingProjectOrchestrator` class referenced above. That is, I didn't want that class to have to worry about Slick types in it's business logic, which would also force leakage of those types to the related tests for mocking purposes (or force the spinup of a full mysql connection).

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 2]
```diff
@@ -523,24 +523,12 @@ class UserService(protected val userInfo: UserInfo,
   def startBillingProjectCreation(createProjectRequest: CreateRawlsBillingProjectFullRequest): Future[Unit] = {
     for {
       _ <- validateV1CreateProjectRequest(createProjectRequest)
-      _ <- checkServicePerimeterAccess(createProjectRequest.servicePerimeter)
+      _ <- BillingProjectOrchestrator.checkServicePerimeterAccess(samDAO, createProjectRequest.servicePerimeter, userInfo)
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala]: The V1 endpoints are still in use and rely on the service perimeter access checks. I've refactored to a method on the companion class so both code paths for reuse.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProjectOrchestrator.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets.UTF_8
+import scala.concurrent.{ExecutionContext, Future}
+
+
+/**
+ * Knows how to provision billing projects with external cloud providers
+ */
+class BillingProjectOrchestrator(userInfo: UserInfo, samDAO: SamDAO, gcsDAO: GoogleServicesDAO, billingRepository: BillingRepository)(implicit val executionContext: ExecutionContext) extends StringValidationUtils {
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingProjectOrchestrator.scala:19]: This class is almost a pure extraction of the v2 billing project code from UserService to get improved (and easier to obtain) coverage on the creation code as I look to further refactor and modify. This [PR](https://github.com/broadinstitute/rawls/pull/1961 ) will introduce the Azure+BPM code path and related API client calls, etc.

**Review by blakery — APPROVED** [review]
Good call on keeping the steps in each individual PR small.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 2 above)
Comment (marctalbott) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala]: There's also a [ServicePerimeterService](https://github.com/broadinstitute/rawls/blob/develop/core/src/main/scala/org/broadinstitute/dsde/rawls/serviceperimeter/ServicePerimeterService.scala) that could be a home for this

**Review by marctalbott — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 2 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala]: @marctalbott Excellent, I did not know that existed, I'll look into it.

**Pull request merged**